k2/csrc/ragged_ops.cu

```diff
@@ -1249,14 +1250,66 @@ std::vector<RaggedShape> GetPrefixes(RaggedShape &src,
 
 Ragged<int32_t> AddSuffixToRagged(Ragged<int32_t> &src,
                                   const Array1<int32_t> &suffix) {
-  K2_LOG(FATAL) << "Not Implemented!";
-  return Ragged<int32_t>();
+  NVTX_RANGE(K2_FUNC);
+  int32_t num_axes = src.NumAxes();
+  K2_CHECK_GE(num_axes, 2);
+  K2_CHECK_EQ(suffix.Dim(), src.TotSize(num_axes - 2));
+  ContextPtr &c = src.Context();
+  Array1<int32_t> dst_values(c, src.NumElements() + suffix.Dim());
+  int32_t *dst_values_data = dst_values.Data();
+  // "row_splits1" and "row_ids1" below are actually on the last axis. We name
+  // them with "1" so that we can use "idx01" and "idx0" for those indexes in
+  // lambda, following the naming convention explained in k2/csrc/utils.h
+  const int32_t *src_row_splits1_data = src.RowSplits(num_axes - 1).Data(),
+                *src_row_ids1_data = src.RowIds(num_axes - 1).Data(),
+                *src_values_data = src.values.Data(),
+                *suffix_data = suffix.Data();
+
+  K2_EVAL(
+      c, src.NumElements(), lambda_copy_src, (int32_t idx01)->void {
+        int32_t dst_idx01 = idx01 + src_row_ids1_data[idx01];
+        dst_values_data[dst_idx01] = src_values_data[idx01];
+      });
+  K2_EVAL(
+       c, suffix.Dim(), lambda_copy_suffix, (int32_t idx0)->void {
+         int32_t dst_idx01 = src_row_splits1_data[idx0 + 1] + idx0;
+         dst_values_data[dst_idx01] = suffix_data[idx0];
+     });
+  RaggedShape dst_shape = ChangeSublistSize(src.shape, 1);
+
+  return Ragged<int32_t>(dst_shape, dst_values);
 }
 
 Ragged<int32_t> AddPrefixToRagged(Ragged<int32_t> &src,
                                   const Array1<int32_t> &prefix) {
-  K2_LOG(FATAL) << "Not Implemented!";
-  return Ragged<int32_t>();
+  NVTX_RANGE(K2_FUNC);
+  int32_t num_axes = src.NumAxes();
+  K2_CHECK_GE(num_axes, 2);
+  K2_CHECK_EQ(prefix.Dim(), src.TotSize(num_axes - 2));
+  ContextPtr &c = src.Context();
+  Array1<int32_t> dst_values(c, src.NumElements() + prefix.Dim());
+  int32_t *dst_values_data = dst_values.Data();
+  // "row_splits1" and "row_ids1" below are actually on the last axis. We name
+  // them with "1" so that we can use "idx01" and "idx0" for those indexes in
+  // lambda, following the naming convention explained in k2/csrc/utils.h
+  const int32_t *src_row_splits1_data = src.RowSplits(num_axes - 1).Data(),
+                *src_row_ids1_data = src.RowIds(num_axes - 1).Data(),
+                *src_values_data = src.values.Data(),
+                *prefix_data = prefix.Data();
+
+  K2_EVAL(
```

Comment:
@danpovey do we need to wrap those two kernels with `ParallelRunner` even if `ParallelRunner` does nothing currently because of cuda record event issue? or merge them into one as I suggested before https://github.com/k2-fsa/k2/pull/453#issuecomment-739109372

Reply:
Mm.  I think it would be better to use a single kernel like you suggested.
The time CUDA would take to deal with these events, or to have sequential kernels run, would likely be much greater than the one or two instructions of extra latency that we'd get in your kernel that has an if-statement.